flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryRow.java

```diff
@@ -0,0 +1,289 @@
+ * used for null tracking and is aligned to 8-byte word boundaries. `Field values` holds
+ * fixed-length primitive types and variable-length values which can be stored in 8 bytes inside.
+ * If it do not fit the variable-length field, then store the length and offset of variable-length
+ * part.
+ *
+ * <p>Fixed-length part will certainly fall into a MemorySegment, which will speed up the read
+ * and write of field. During the write phase, if the target memory segment has less space than
+ * fixed length part size, we will skip the space. So the number of fields in a single Row cannot
+ * exceed the capacity of a single MemorySegment, if there are too many fields, we suggest that
+ * user set a bigger pageSize of MemorySegment.
+ *
+ * <p>Variable-length part may fall into multiple MemorySegments.
+ *
+ * <p>{@code BinaryRow} are influenced by Apache Spark UnsafeRow in project tungsten.
+ * The difference is that BinaryRow is placed on a discontinuous memory, and the variable length
+ * type can also be placed on a fixed length area (If it's short enough).
+ */
+public final class BinaryRow extends BinaryFormat<Object> implements BaseRow {
+
+	public static final boolean LITTLE_ENDIAN = (ByteOrder.nativeOrder() == ByteOrder.LITTLE_ENDIAN);
+	public static final long FIRST_BYTE_ZERO = LITTLE_ENDIAN ? 0xFFF0 : 0x0FFF;
+
+	public static int calculateBitSetWidthInBytes(int arity) {
+		// add 8 bit header
+		return ((arity + 63 + 8) / 64) * 8;
+	}
+
+	private final int arity;
+	private final int nullBitsSizeInBytes;
+
+	public BinaryRow(int arity) {
+		checkArgument(arity >= 0);
+		this.arity = arity;
+		this.nullBitsSizeInBytes = calculateBitSetWidthInBytes(arity);
+	}
+
+	private int getFieldOffset(int pos) {
+		return offset + nullBitsSizeInBytes + pos * 8;
+	}
+
+	private void assertIndexIsValid(int index) {
+		assert index >= 0 : "index (" + index + ") should >= 0";
+		assert index < arity : "index (" + index + ") should < " + arity;
+	}
+
+	public int getFixedLengthPartSize() {
+		return nullBitsSizeInBytes + 8 * arity;
+	}
+
+	@Override
+	public int getArity() {
+		return arity;
+	}
+
+	@Override
+	public byte getHeader() {
+		// first nullBitsSizeInBytes byte is header.
+		return segments[0].get(offset);
+	}
+
+	@Override
+	public void setHeader(byte header) {
+		segments[0].put(offset, header);
+	}
+
+	public void pointTo(MemorySegment segment, int offset, int sizeInBytes) {
+		this.segments = new MemorySegment[] {segment};
+		this.offset = offset;
+		this.sizeInBytes = sizeInBytes;
+	}
+
+	public void pointTo(MemorySegment[] segments, int offset, int sizeInBytes) {
+		this.segments = segments;
+		this.offset = offset;
+		this.sizeInBytes = sizeInBytes;
+	}
+
+	public void setTotalSize(int sizeInBytes) {
+		this.sizeInBytes = sizeInBytes;
+	}
+
+	private void setNotNullAt(int i) {
+		assertIndexIsValid(i);
+		// need add header 8 bit.
+		SegmentsUtil.bitUnSet(segments[0], offset, i + 8);
```

Comment:
How about we turn the magic number 8 to something like HEAER_LENGTH ?

Reply:
OK